presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizer.java

```diff
@@ -0,0 +1,235 @@
+    private final CounterStat failureCount = new CounterStat();
+
+    @Inject
+    public ShardOrganizer(@ForMetadata IDBI dbi, ShardManager shardManager, ShardCompactor compactor, StorageManagerConfig config)
+    {
+        this(dbi, shardManager, compactor, config.getCompactionThreads());
+    }
+
+    public ShardOrganizer(@ForMetadata IDBI dbi, ShardManager shardManager, ShardCompactor compactor, int threads)
+    {
+        this.metadataDao = onDemandDao(dbi, MetadataDao.class);
+        this.shardManager = requireNonNull(shardManager, "shardManager is null");
+        this.compactor = requireNonNull(compactor, "compactor is null");
+
+        this.executorService = newFixedThreadPool(threads, daemonThreadsNamed("shard-organizer-%s"));
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        driverService.submit((Runnable) () -> {
+            while (!Thread.currentThread().isInterrupted() && !shutdown.get()) {
+                try {
+                    OrganizationSet compactionSet = queue.take();
+                    runAsync(new OrganizationJob(compactionSet), executorService)
+                            .whenComplete((none, throwable) -> {
+                                if (throwable == null) {
+                                    successCount.update(1);
+                                }
+                                else {
+                                    log.warn(throwable, "Error in running organization job");
+                                    failureCount.update(1);
+                                }
+                            });
+                }
+                catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        });
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        shutdown.set(true);
+        driverService.shutdown();
+        executorService.shutdown();
+    }
+
+    void enqueue(OrganizationSet organizationSet)
+    {
+        shardsInProgress.addAll(organizationSet.getShards());
+        queue.add(organizationSet);
+    }
+
+    boolean inProgress(UUID shardUuid)
+    {
+        return shardsInProgress.contains(shardUuid);
+    }
+
+    public class OrganizationJob
+            implements Runnable
+    {
+        private final OrganizationSet organizationSet;
+
+        public OrganizationJob(OrganizationSet organizationSet)
+        {
+            this.organizationSet = requireNonNull(organizationSet, "organizationSet is null");
+        }
+
+        @Override
+        public void run()
+        {
+            try {
+                runJob(organizationSet.getTableId(), organizationSet.getBucketNumber(), organizationSet.getShards());
+            }
+            catch (IOException e) {
+                throw Throwables.propagate(e);
+            }
+            finally {
+                shardsInProgress.removeAll(organizationSet.getShards());
```

Comment:
`shardsInProgress` is a concern of the driver, not the job. Move this to the `whenComplete` handler